### PR TITLE
Add support to build and link common code in tests

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Add support to build and link to test suites common code located in `tests` folder.
    
The common code is built into an mbedtests library and linked to the generated test suite code along with the mbedtls library.
    
To start with, move helper functions (helpers.function) into an helpers.c file and declare the interface it provides to the test suites into its associated helpers.h file.
    
The commit just dispatch the code and comments of helpers.function into helpers.c and helpers.h. There is no code modification and just a few comment adjustments.

This implements partially #3325.

## Status
READY

## Requires Backporting
I'd say so but not entirely sure.
Which branch: 2.7 and 2.16.

## Todos
- [ ] Backported

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Please break up the commit “tests: Add support to build and link common code” into several pieces. It's hard to follow with so many things going on. Do not move code and do other things in the same commit. When moving code to a new file, make a commit that creates a new file that is functionally empty (only headers, include guards, etc.), add it to the build scripts in the same commit or subsequent commits depending on how complex that is, and then make a commit that does nothing but move code from one file to another (or from/to multiple files if it isn't too messy).

**Comment by ronald-cron-arm** [comment]
Yes, no problem, I will.

File: tests/include/helpers.h [hunk 1]
```diff
@@ -69,6 +69,382 @@ typedef UINT32 uint32_t;
...
+    ( !defined(MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES) && \
+      ( !defined(MBEDTLS_NO_PLATFORM_ENTROPY)  ||     \
+         defined(MBEDTLS_HAVEGE_C)             ||     \
+         defined(MBEDTLS_ENTROPY_HARDWARE_ALT) ||     \
+         defined(ENTROPY_NV_SEED) ) )
+#define ENTROPY_HAVE_STRONG
+#endif
+
+/*----------------------------------------------------------------------------*/
+/* Status and error constants */
```
Comment (gilles-peskine-arm) [inline tests/include/helpers.h:87]: The dispatch status values are an internal detail of the toplevel loop of the test code, not for use in test code. Since this PR is precisely about moving code around, I think it would be better to move this directly to a separate header file (even if this header would currently be included in the same places). But this isn't a blocker.

This also applies to the definition of `test_info_t` and related types.

File: tests/include/helpers.h [hunk 2]
```diff
@@ -69,6 +69,382 @@ typedef UINT32 uint32_t;
...
+
+/* Indicates whether we expect mbedtls_entropy_init
+ * to initialize some strong entropy source. */
+#if defined(MBEDTLS_TEST_NULL_ENTROPY) ||             \
+    ( !defined(MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES) && \
+      ( !defined(MBEDTLS_NO_PLATFORM_ENTROPY)  ||     \
+         defined(MBEDTLS_HAVEGE_C)             ||     \
+         defined(MBEDTLS_ENTROPY_HARDWARE_ALT) ||     \
+         defined(ENTROPY_NV_SEED) ) )
+#define ENTROPY_HAVE_STRONG
```
Comment (gilles-peskine-arm) [inline tests/include/helpers.h:83]: It would be nice to update the comment in `library/entropy.c` to point to the new location here.

File: tests/src/Makefile [hunk 3]
```diff
@@ -0,0 +1,68 @@
+CFLAGS	?= -O2
```
Comment (gilles-peskine-arm) [inline tests/src/Makefile:1]: `tests/src/Makefile` repeats a lot of definitions that are in `tests/Makefile`. I don't see why we need to have a separate makefile in `tests/src`. [Recursive make is an antipattern](http://www.real-linux.org.uk/recursivemake.pdf) . Why not put the relevant targets in `tests/Makefile`, the same way `programs/Makefile` has targets for all the program subdirectories?

File: tests/include/helpers.h (hunk 1 above)
Comment (ronald-cron-arm) [reply tests/include/helpers.h]: I don't move this good in the new version of the PR.

File: tests/include/helpers.h (hunk 2 above)
Comment (ronald-cron-arm) [reply tests/include/helpers.h]: I don't move this code in the new version of the PR.

File: tests/src/Makefile (hunk 3 above)
Comment (ronald-cron-arm) [reply tests/src/Makefile]: `tests/src/Makefile` has been removed.

**Comment by ronald-cron-arm** [comment]
Following @gilles-peskine-arm comments, I've reworked significantly the PR. The previous version can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/tests-common-code-1).

The PR doesn't move all helpers.function code to new files anymore. It moves only the code that is not related to the unit test code execution machinery. That way the PR is more focused on its primary goal which is to be able to build a test library and link it to test suite executables.  I am happy to consider moving further code from helpers.function to .c and .h files in tests/src and tests/include but in another PR.

Other major changes:
- Test header files are located in tests/include/test and not in include directly anymore. As the test include path in the build systems is still tests/include, this implies that test headers are included using the `test/*` pattern. This makes explicit the scope of test headers and aligns with the `mbedtls/*` and `psa/*` patterns.
- To minimize name collision risk, I've prefixed the test global functions introduced by this PR with `mbedtls_test_` and fixed code alignment issues in test suites `.function` files due to renaming.

File: ChangeLog.d/tests-common-code.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+   * Test specific header files are now located in a dedicated 'include/test'
+     folder in 'tests' folder, not anymore directly in the 'tests' folder.
+   * An 'src' folder has been created in 'tests' folder to contain common test
+     code. The build systems have been modified to build the C files in that
+     'src' folder into an 'mbedtests' library that is linked to the test
+     suites.
+   * Test helper macros and functions from helpers.function, that are not
+     related to the unit test execution machinery have been moved to .c and
+     .h files located respectively in 'tests/src' and 'tests/include/test'
+     folders.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/tests-common-code.txt:11]: I think the changelog entry is a bit too verbose, and all the changes should be in a single entry. The audience is people who build the tests, not people who maintain the test code. What we need to convey is that if you were using your own build instructions to build the tests, you need to update them. Suggestion:

> The unit tests now rely on header files in tests/include/test and source files in tests/src. When building with make or cmake, the files in tests/src are compiled to a libmbedtests library which is linked statically into each test executable (even when the product libraries are linked dynamically).

File: tests/Makefile [hunk 5]
```diff
@@ -21,11 +30,13 @@ LOCAL_CFLAGS+=$(THIRDPARTY_INCLUDES)
...
 
 ifndef SHARED
-DEP=../library/libmbedcrypto.a ../library/libmbedx509.a ../library/libmbedtls.a
+MBEDLIBS=../library/libmbedcrypto.a ../library/libmbedx509.a ../library/libmbedtls.a
 else
-DEP=../library/libmbedcrypto.$(DLEXT) ../library/libmbedx509.$(DLEXT) ../library/libmbedtls.$(DLEXT)
+MBEDLIBS=../library/libmbedcrypto.$(DLEXT) ../library/libmbedx509.$(DLEXT) ../library/libmbedtls.$(DLEXT)
 endif
 
+MBEDTESTSLIB=./src/libmbedtests.a
```
Comment (gilles-peskine-arm) [inline tests/Makefile:38]: It would be good to add a comment stating that this is deliberately linked statically even under `SHARED`. Likewise in `CMakeLists.txt`.

Comment (gilles-peskine-arm) [inline tests/Makefile:38]: Nitpick: why the `./` ?

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -1,15 +1,31 @@
...
 
+# For testing purposes, the Mbed TLS libraries are expected to rely on common
+# test code, for example, of the testing of the crypto library backed-up by
+# PSA drivers.
+# This dependency cannot be defined as part of the library build system as we
+# want libraries to be completely independant of test code and test build
+# system. For the unit tests, the dependencies are thus defined here.
+# We specify (through the target_link_libraries command) each test executable to
+# link against mbedtls, mbedx509, mbedcrypto and mbedtests in that order. The
+# order is important as:
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/CMakeLists.txt:16]: Well, I am not completely sure to understand your concern. I've tried to explain the need for ordering but failed to make it clear enough it seems.
The root problem here is that it seems there is no "linker independent" proper solution, thus basically no proper solution, to the problem of mutual/cyclic dependencies between libraries. It sounds a bit incredible but that the way it is .
The usual nasty way of dealing with mutual/cyclic dependencies is to repeat several time in an "appropriate" order the libraries in the link command line like: -llibA -llibB -llibA -llibB ... and you end up with the above kind of obscure comments to explain this nightmare.
Eventually, I think I will move to a list of objects for the test common code. This will avoid the library cyclic dependencies issue and should be good enough at least for the time being. Maybe, that we will feel the need for proper libraries for test drivers but we will see at that point.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/CMakeLists.txt:16]: I know that there's no portable way to deal with cyclic dependencies. But the commit message and the comment together seem contradictory: on the one hand there are cyclic dependencies, but on the other hand the link needs to happen in this particular order because the dependencies are in this order.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply tests/CMakeLists.txt:16]: And indeed this is a problem that I hadn't thought of, and I agree to do whatever is simple and works now, and worry later if it isn't enough. I just want it to be clear what we're doing now and what assumptions we're making.

File: tests/Makefile (hunk 5 above)
Comment (ronald-cron-arm) [reply tests/Makefile]: This code has been removed.

Comment (ronald-cron-arm) [reply tests/Makefile]: This code has been removed.

File: ChangeLog.d/tests-common-code.txt (hunk 4 above)
Comment (ronald-cron-arm) [reply ChangeLog.d/tests-common-code.txt]: I have adapted your proposal to the fact that there is no libmbedtests library anymore.

**Comment by ronald-cron-arm** [comment]
The version before the last force-push can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/tests-common-code-2).
The mbedtests library has been been removed to avoid having to handle its mutual dependency with the Mbed libraries. The test common C files under tests/src are now compiled and the resulting objects directly linked to test suite executables.

**Comment by mpg** [comment]
From the PR description:

> Requires Backporting
> I'd say so but not entirely sure.

I don't think this is an appropriate change for the LTS branches, as it changes how tests are built, and as Gilles said, it people are not using our provided Makefiles/CMakeLists.txt, the changes might break their build scripts. I think it's an acceptable risk for development, but perhaps less so for the LTS branches.

**Comment by mpg** [comment]
I think there's just one question to resolve before merging this: do we want to backport it? @gilles-peskine-arm wdyt?
